src/mumble/AudioInput.cpp

```diff
@@ -27,6 +27,102 @@ extern "C" {
 }
 #endif
 
+void Resynchronizer::addMic(short *mic)
+{
+	bool drop = false;
+	{
+		std::unique_lock<std::mutex> l(m);
+		micQueue.push_back(mic);
+		switch(state)
+		{
+			case S0:  state = S1a; break;
+			case S1a: state = S2;  break;
+			case S1b: state = S2;  break;
+			case S2:  state = S3;  break;
+			case S3:  state = S4a; break;
+			case S4a: state = S5;  break;
+			case S4b: drop = true; break;            
+			case S5:  drop = true; break;
+		}
+		if(drop)
+		{
+			delete[] micQueue.front();
+			micQueue.pop_front();
+		}
+	}
+	if(bDebugPrintQueue)
+	{
+		if(drop) printf("dropped microphone chunk due to overflow\n");
+		printQueue('+');
+	}
+}
+
+AudioChunk Resynchronizer::addSpeaker(short *speaker)
+{
+	AudioChunk result;
+	bool drop = false;
+	{
+		std::unique_lock<std::mutex> l(m);
+		switch(state)
+		{
+			case S0:  drop = true; break;
+			case S1a: drop = true; break;
+			case S1b: state = S0;  break;
+			case S2:  state = S1b; break;
+			case S3:  state = S2;  break;
+			case S4a: state = S3;  break;
+			case S4b: state = S3;  break;
+			case S5:  state = S4b; break;
+		}
+		if(drop == false)
+		{
+			result = AudioChunk(micQueue.front(), speaker);
+			micQueue.pop_front();
+		}
+	}
+	if(drop) delete[] speaker;
+	if(bDebugPrintQueue)
+	{
+		if(drop) printf("dropped speaker chunk due to underflow\n");
+		printQueue('-');
+	}
+	return result;
+}
+
+void Resynchronizer::reset()
+{
+	if(bDebugPrintQueue) printf("resetting echo queue\n");
+	std::unique_lock<std::mutex> l(m);
+	state = S0;
+	while(micQueue.empty() == false)
+	{
+		delete[] micQueue.front();
+		micQueue.pop_front();
+	}
+}
+
+Resynchronizer::~Resynchronizer()
+{
+	reset();
+}
+
+void Resynchronizer::printQueue(char who)
+{
+	unsigned int mic;
+	{
+		std::unique_lock<std::mutex> l(m);
+		mic = static_cast<unsigned int>(micQueue.size());
+	}
+	std::string line;
+	line.reserve(32);
+	line += who;
+	line += " Echo queue [";
+	for(unsigned int i = 0; i < 5; i++) line += i < mic ? '#' : ' ';
+	line += "]\r";
+	printf("%s",line.c_str());
```

Comment:
`printf`-> `qWarning`
I think even with the parameter you can just replace the function name. In case that causes errors, here's the doc for `qWarning`: https://doc.qt.io/qt-5/qtglobal.html#qWarning